lib/support.nix

```diff
@@ -97,25 +98,33 @@ rec {
       excluded = map sanitizeRelativePath (manifest.workspace.exclude or []);
       members = subtractLists excluded selected;
 
-      lock = fromTOML (readFile (src + "/Cargo.lock"));
+      lockSrc = if lockPath == null then (src + "/Cargo.lock") else lockPath;
+      lock = fromTOML (readFile lockSrc);
       # We don't distinguish between v1 and v2. But v3 is different from both.
       lockVersionSet = { lockVersion = lock.version or 2; };
 
       localSrcInfos =
-        listToAttrs
-        (map (relativePath:
-          let
-            memberRoot = src + ("/" + relativePath);
-            memberManifest = fromTOML (readFile (memberRoot + "/Cargo.toml")) // lockVersionSet;
-          in {
-            name = toPkgId memberManifest.package;
-            value = mkPkgInfoFromCargoToml memberManifest memberRoot;
-          }
-          ) (if manifest ? workspace then members else [ "" ]));
+        let
+          workspace_members = (if manifest ? workspace then members else []);
+          root_package = (if manifest ? package then [ "" ] else []);
+          main-workspace = if manifest ? "workspace" then manifest.workspace else {};
+          maybe-crates = lib.attrNames (lib.filterAttrs (path: type: type == "directory") (builtins.readDir src));
```

Comment:
I think I changed this because they were using crates from the sub directory without listing them in the `members` section (I think `cargo` did this). The proper way would be to crawl the Cargo toml and figure out which of them should be included, but AFAIK, all this does is list these crates as "possible crates to include", so it shouldn't make any difference if they aren't imported.

Reply:
> I think I changed this because they were using crates from the sub directory without listing them in the `members` section (I think `cargo` did this).

Yes and no. According to [cargo reference](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields),

> All [path dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies) residing in the workspace directory automatically become members. Additional members can be listed with the members key, ...

If one sub directory is not included by `path` dependencies, it's a standalone package or another workspace, which needs their own `Cargo.lock`. It's an uncommon usage, but still possible.

> The proper way would be to crawl the Cargo toml and figure out which of them should be included, but AFAIK, all this does is list these crates as "possible crates to include", so it shouldn't make any difference if they aren't imported.

The difference is non-members should not be resolved and their features should not be merged into the current workspace. Eg. if workspace members `a` and `b` both depend on `c` but with different features, their features will get merged, and both `a` and `b` will be built on the `c` with union of all features enabled (thus they can interoperate). If `b` is not a member, this never happen, and they can be built on different feature set of `c`.